REGRESSION(264685@main): Introduced page load time regression

pvollan opened the pull request:
#### 695ee5dcb9e5d659b4b7def7736630fdbf109276
<pre>
REGRESSION(264685@main): Introduced page load time regression
<a href="https://bugs.webkit.org/show_bug.cgi?id=258887">https://bugs.webkit.org/show_bug.cgi?id=258887</a>
rdar://111788974

Reviewed by Brent Fulgham.

When the log daemon is blocked, the WebContent process will have to read log preferences directly
from plist files on disk, which can slow down page loading. Address this issue by prewarming some
log objects.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::prewarmLogs):
(WebKit::WebProcess::platformInitializeProcess):

Canonical link: <a href="https://commits.webkit.org/265807@main">https://commits.webkit.org/265807@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0cc8a056c219780ab93c33bcd2b346dbda172b7a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11867 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12066 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12439 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13512 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11304 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14455 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12047 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14162 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12031 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12856 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13934 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10148 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17904 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11226 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10929 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14098 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11342 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9378 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10517 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2920 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14800 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11198 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a6195610c0880082ea96ed8570347868c4487d36. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/15569

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0cc8a056c219780ab93c33bcd2b346dbda172b7a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/15569

brentfulgham commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -714,6 +714,47 @@ NEVER_INLINE NO_RETURN_DUE_TO_CRASH static void deliberateCrashForTesting()
+    for (auto& log : logs) {
+        auto logHandle = os_log_create(log.first, log.second);
+        bool enabled = os_log_type_enabled(logHandle, OS_LOG_TYPE_ERROR);
+        UNUSED_PARAM(enabled);
```
Should we take any action if `os_log_type_enabled` fails? I guess not, since if we can't create a log element, there's no way to document that.

pvollan replied:
> Should we take any action if `os_log_type_enabled` fails? I guess not, since if we can't create a log element, there's no way to document that.

Yes, I agree.

Thanks for reviewing!

brentfulgham reviewed (approved):
Looks good. I wish `std::array` didn't require you to hard-code the expected size!.

cdumez commented:
> Looks good. I wish `std::array` didn't require you to hard-code the expected size!.

It doesn't require you:
```
[std::array](http://en.cppreference.com/w/cpp/container/array) a{1, 2, 3, 5, x}; // OK, creates std::array<int, 5>
```

aproskuryakov commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -714,6 +714,47 @@ NEVER_INLINE NO_RETURN_DUE_TO_CRASH static void deliberateCrashForTesting()
 #if ENABLE(LOGD_BLOCKING_IN_WEBCONTENT)
+static void prewarmLogs()
+{
+    static std::array<std::pair<const char*, const char*>, 29> logs { {
```
It it feasible at all to keep this list up to date in the future? Seems like we'd be getting performance regressions whenever something else needs to be added, and investigating those is time consuming and unreliable.

Perhaps I'm not aware of some reason to block logging, but perhaps now is the time to reconsider the plan, and to just unblock it? Logs are useful.

pvollan replied:
> It it feasible at all to keep this list up to date in the future? Seems like we'd be getting performance regressions whenever something else needs to be added, and investigating those is time consuming and unreliable.
> 
> Perhaps I'm not aware of some reason to block logging, but perhaps now is the time to reconsider the plan, and to just unblock it? Logs are useful.



> It it feasible at all to keep this list up to date in the future? Seems like we'd be getting performance regressions whenever something else needs to be added, and investigating those is time consuming and unreliable.
> 
> Perhaps I'm not aware of some reason to block logging, but perhaps now is the time to reconsider the plan, and to just unblock it? Logs are useful.

That is a good point! I think we can replace the container manager entries with a call to topPrivatelyControlledDomain(), since that is the call that ends up creating these logs, since the CFNetwork API used in that function contains container manager calls. I will look into that in a follow-up patch. We are blocking the log daemon, but not the WebContent logs themselves. These are now logged via the Networking process.

Thanks for reviewing!

pvollan commented:
> > Looks good. I wish `std::array` didn't require you to hard-code the expected size!.
> 
> It doesn't require you:
> 
> ```
> [std::array](http://en.cppreference.com/w/cpp/container/array) a{1, 2, 3, 5, x}; // OK, creates std::array<int, 5>
> ```

I tried without the size, but it gave me a compile error about missing arguments.

Thanks for reviewing!

pvollan committed:
REGRESSION(264685@main): Introduced page load time regression
https://bugs.webkit.org/show_bug.cgi?id=258887
rdar://111788974

Reviewed by Brent Fulgham.

When the log daemon is blocked, the WebContent process will have to read log preferences directly
from plist files on disk, which can slow down page loading. Address this issue by prewarming some
log objects.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::prewarmLogs):
(WebKit::WebProcess::platformInitializeProcess):

Canonical link: https://commits.webkit.org/265807@main

webkit-commit-queue commented:
Committed 265807@main (695ee5dcb9e5): <https://commits.webkit.org/265807@main>

Reviewed commits have been landed. Closing PR #15569 and removing active labels.
